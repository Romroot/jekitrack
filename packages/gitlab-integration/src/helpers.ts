/* eslint-disable camelcase */
import format from 'date-fns/format'
import { ReportGroupType } from './types'

export const extractTicketNumber = (str: string) => {
  const result = str.match(/GENESIS[-_][0-9]+/)?.[0]
  return result?.replace('_', '-')
}

export const getTicketNumber = (event: any) => {
  const items = [event?.push_data?.commit_title, event?.push_data?.ref, event?.target_title].filter(Boolean)
  const tStr = items.find(extractTicketNumber)

  if (!tStr) {
    const isPoc = items.some((e) => e?.toString()?.toUpperCase()?.includes('POC'))
    if (isPoc || process.env.NODE_ENV !== 'development') {
      return null
    }
    throw new Error(`Can not find ticket for ${JSON.stringify(event)} `)
  }
  return extractTicketNumber(tStr)
}

export const getEventDescription = (event: any) => {
  return (
    event?.push_data?.commit_title ||
    (event?.action_name?.includes('commented') && `Commented at: ${event?.target_title}`) ||
    (event?.action_name?.includes('action_name') && `Opened MR: ${event.target_title}`) ||
    'Unknown action'
  )
}

export const groupEvents = (eventList: any[]): ReportGroupType => {
  const events = eventList.filter((e) => !(e.action_name?.includes('pushed') && e?.push_data?.ref === 'master'))

  const result: any = {}
  events.forEach((e: any) => {
    const dateKey = format(new Date(e?.created_at ?? e?.note?.created_at), 'yyyy-MM-dd')
    result[dateKey] = result[dateKey] ?? []
    const ticket = getTicketNumber(e)
    if (!ticket) {
      return
    }
    let data: any = result[dateKey]?.find((l: any) => l.ticket === ticket)
    const actionName = e.action_name
    const eventDescription = getEventDescription(e)
    if (data) {
      data.actions[actionName] = [...(data.actions[actionName] || []), eventDescription]
    } else {
      data = { ticket, actions: { [actionName]: [eventDescription] } }
      result[dateKey].push(data)
    }
  })

  return result
}
