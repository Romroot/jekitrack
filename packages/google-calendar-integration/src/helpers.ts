import format from 'date-fns/format'
import differenceInMinutes from 'date-fns/differenceInMinutes'
import { ReportGroupType } from './types'

export const groupEvents = (times: any[]): ReportGroupType => {
  const result = times.reduce((acc, e) => {
    const key = format(new Date(e?.start), 'yyyy-MM-dd')
    return { ...acc, [key]: (acc[key] || 0) + differenceInMinutes(new Date(e?.end), new Date(e?.start)) }
  }, {})

  Object.keys(result).forEach((k) => {
    result[k] = [
      {
        ticket: 'ITL-3',
        time: Math.round(result[k] / 60) || 0.5,
        actions: ['meeting'],
      },
    ]
  })

  return result
}
