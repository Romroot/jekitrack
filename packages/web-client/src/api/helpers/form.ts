import { FORM_ERROR } from 'final-form'

export const withFormRequest = (fn: any) => async (...args: any[]) => {
  try {
    await fn(...args)
    return undefined
  } catch (e) {
    return { [FORM_ERROR]: e.response?.data?.message ?? e?.message ?? 'Failed to perfom action' }
  }
}
