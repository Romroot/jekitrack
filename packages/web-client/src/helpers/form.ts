export const extractFieldError = (meta: any, submitting = false) => {
  if (!meta || submitting) {
    return ''
  }

  const { data, error, submitError, touched, dirty, dirtySinceLastSubmit, pristine, active, submitFailed } = meta

  const showSubmitError = !dirtySinceLastSubmit
  const showDataError = touched && !dirty
  const showSynError = submitFailed
    ? pristine || !active // after form was submitted and failed
    : !active && touched // before form was submitted

  return (showSubmitError && submitError) || (showDataError && data && data.error) || (showSynError && error) || ''
}

// We show form submit error when: fields are pristine since last submit and form is submitting
export const extractFormError = ({ dirtySinceLastSubmit, submitError, submitting }: any) => {
  return !dirtySinceLastSubmit && !submitting && submitError
}

// External submit form
// https://final-form.org/docs/react-final-form/faq
export const submitForm = (formId: string) => {
  return document.getElementById(formId)!.dispatchEvent(new Event('submit', { cancelable: true }))
}
