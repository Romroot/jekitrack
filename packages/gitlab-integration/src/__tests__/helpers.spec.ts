import * as fixtures from './fixtures'
import * as helpers from '../helpers'

describe('[Helpers]', () => {
  describe('extractTicketNumber()', () => {
    it('should work', () => {
      expect(helpers.extractTicketNumber('[GENESIS-46426] Added form library.')).toBe('GENESIS-46426')
    })
  })

  describe('groupEvents', () => {
    it('should group by dates', () => {
      expect(Object.keys(helpers.groupEvents(fixtures.demoEvents0))).toEqual([
        '2020-07-10',
        '2020-07-09',
        '2020-07-08',
        '2020-07-07',
        '2020-07-06',
      ])
    })

    it('should extract actions with messages', () => {
      expect(helpers.groupEvents(fixtures.demoEvents0)).toMatchSnapshot()
    })
  })
})
