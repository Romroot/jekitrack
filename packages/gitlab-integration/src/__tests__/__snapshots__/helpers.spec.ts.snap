// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Helpers] groupEvents should extract actions with messages 1`] = `
Object {
  "2020-07-06": Array [
    Object {
      "actions": Object {
        "pushed to": Array [
          "[GENESIS-43100] Small fix.",
          "[GENESIS-43100] Small change.",
        ],
      },
      "ticket": "GENESIS-43100",
    },
    Object {
      "actions": Object {
        "opened": Array [
          "Unknown action",
        ],
        "pushed new": Array [
          "[GENESIS-46425] Adding base form library.",
        ],
        "pushed to": Array [
          "[GENESIS-46425] Small change.",
        ],
      },
      "ticket": "GENESIS-46425",
    },
  ],
  "2020-07-07": Array [
    Object {
      "actions": Object {
        "commented on": Array [
          "Commented at: GENESIS-46425 \\"Add form module\\"",
          "Commented at: GENESIS-46425 \\"Add form module\\"",
        ],
        "pushed to": Array [
          "[GENESIS-46425] Updating package.json.",
          "[GENESIS-46425] Fixing linter.",
          "Merge master into GENESIS-46425_add_form_module",
          "[GENESIS-46425] Updated README.md",
          "[GENESIS-46425] Rolled back customer indv change.",
          "[GENESIS-46425] Base form library package.",
        ],
      },
      "ticket": "GENESIS-46425",
    },
  ],
  "2020-07-08": Array [
    Object {
      "actions": Object {
        "accepted": Array [
          "Unknown action",
        ],
        "deleted": Array [
          "Unknown action",
        ],
        "pushed to": Array [
          "[GENESIS-46425] Temp commented BR tests.",
        ],
      },
      "ticket": "GENESIS-46425",
    },
  ],
  "2020-07-09": Array [
    Object {
      "actions": Object {
        "commented on": Array [
          "Commented at: GENESIS-46563 \\"Yarn audit job fails\\"",
          "Commented at: GENESIS-46563 \\"Yarn audit job fails\\"",
          "Commented at: GENESIS-46563 \\"Yarn audit job fails\\"",
          "Commented at: GENESIS-46563 \\"Yarn audit job fails\\"",
          "Commented at: GENESIS-46563 \\"Yarn audit job fails\\"",
        ],
      },
      "ticket": "GENESIS-46563",
    },
    Object {
      "actions": Object {
        "commented on": Array [
          "Commented at: [GENESIS-30498] Unit tests and Story for MoneyInput component",
        ],
      },
      "ticket": "GENESIS-30498",
    },
    Object {
      "actions": Object {
        "pushed new": Array [
          "[GENESIS-46425] Added translate, updated hooks/HOCs.",
        ],
      },
      "ticket": "GENESIS-46425",
    },
  ],
  "2020-07-10": Array [
    Object {
      "actions": Object {
        "opened": Array [
          "Unknown action",
        ],
        "pushed to": Array [
          "[GENESIS-46426] Added form library.",
          "[GENESIS-46426] Form library update.",
          "[GENESIS-46426] Working on form library.",
        ],
      },
      "ticket": "GENESIS-46426",
    },
  ],
}
`;
